File: techpath-admin/src/types/training.ts
// Training subsystem types. Mirrors app/schemas/training*.py on the backend.

export type AssetStorageKind = 'inline_text' | 'file' | 'link' | 'structured' | 'bundle';

export type AssetType =
  | 'markdown'
  | 'notes'
  | 'cheat_sheet'
  | 'code_snippet'
  | 'pdf'
  | 'ppt'
  | 'video'
  | 'notebook'
  | 'zip'
  | 'excel'
  | 'csv'
  | 'terminal_recording'
  | 'external_url'
  | 'github_repo'
  | 'youtube'
  | 'quiz'
  | 'assignment'
  | 'lab'
  | 'html_bundle';

export type ContentStatus = 'draft' | 'published' | 'archived';
export type DeliveryMode = 'online' | 'offline' | 'hybrid';
export type SessionStatus = 'scheduled' | 'live' | 'ended' | 'cancelled';

/** Registry entry served by GET /training/asset-types. */
export interface AssetTypeInfo {
  value: AssetType;
  label: string;
  storage_kind: AssetStorageKind;
  max_size_mb: number;
  allowed_content_types: string[];
  allowed_extensions: string[];
}

export interface QuizQuestion {
  question: string;
  options: string[];
  correct_index: number;
  explanation?: string | null;
}

export interface LectureAsset {
  id: number;
  public_id: string;
  title: string;
  asset_type: AssetType;
  description?: string | null;
  body?: string | null;
  media_file_id?: number | null;
  external_url?: string | null;
  config?: Record<string, unknown> | null;
  tags: string[];
  status: ContentStatus;
  is_active: boolean;
  storage_kind: AssetStorageKind;
  /** Resolved server-side from media_file_id — never construct this from config yourself,
   *  Azure Blob storage needs a signed URL the client can't produce. */
  file_url?: string | null;
  /** Also resolved server-side — a browser-side fetch of file_url would be blocked by
   *  the storage provider's CORS policy (Azure Blob grants none by default). */
  csv_preview?: {
    header: string[];
    rows: string[][];
    total_rows: number;
    truncated: boolean;
  } | null;
  created_at: string;
  updated_at: string;
}

export interface AssetUsage {
  module_id: number;
  module_title: string;
  program_id: number;
  program_title: string;
}

export interface ModuleAssetLink {
  id: number;
  asset_id: number;
  display_order: number;
  is_required: boolean;
  notes?: string | null;
  asset: LectureAsset;
}

export interface TrainingModule {
  id: number;
  program_id: number;
  title: string;
  slug: string;
  description?: string | null;
  display_order: number;
  estimated_minutes?: number | null;
  status: ContentStatus;
  asset_count: number;
  created_at: string;
  updated_at: string;
}

export interface TrainingModuleDetail extends TrainingModule {
  assets: ModuleAssetLink[];
}

export interface TrainingProgram {
  id: number;
  title: string;
  slug: string;
  summary?: string | null;
  description?: string | null;
  course_id?: number | null;
  delivery_mode: DeliveryMode;
  level?: string | null;
  duration?: string | null;
  cover_image?: string | null;
  tags: string[];
  status: ContentStatus;
  display_order: number;
  is_active: boolean;
  module_count: number;
  created_at: string;
  updated_at: string;
}

export interface TrainingProgramDetail extends TrainingProgram {
  modules: TrainingModule[];
}

export interface TrainingProgramCreate {
  title: string;
  slug: string;
  summary?: string;
  description?: string;
  course_id?: number | null;
  delivery_mode?: DeliveryMode;
  level?: string;
  duration?: string;
  cover_image?: string;
  tags?: string[];
  status?: ContentStatus;
  display_order?: number;
}

export type TrainingProgramUpdate = Partial<TrainingProgramCreate> & { is_active?: boolean };

export interface TrainingModuleCreate {
  title: string;
  slug: string;
  description?: string;
  display_order?: number;
  estimated_minutes?: number;
  status?: ContentStatus;
}

export type TrainingModuleUpdate = Partial<TrainingModuleCreate>;

export interface ReorderItem {
  id: number;
  display_order: number;
}

// ---------- roster mirror ----------

export interface TrainingBatch {
  id: number;
  external_id: string;
  name: string;
  code?: string | null;
  program_id?: number | null;
  program_title?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  timezone?: string | null;
  schedule?: {
    days?: string[];
    start_time?: string;
    end_time?: string;
    timezone?: string;
  } | null;
  status?: string | null;
  mode?: string | null;
  location?: string | null;
  trainer_email?: string | null;
  trainer_name?: string | null;
  student_count: number;
  course_ref?: string | null;
  synced_at?: string | null;
  external_updated_at?: string | null;
}

export interface TrainingStudent {
  id: number;
  external_id: string;
  name: string;
  email?: string | null;
  phone?: string | null;
  roll_no?: string | null;
  status?: string | null;
  photo_url?: string | null;
  enrolled_on?: string | null;
  synced_at?: string | null;
}

export interface SyncState {
  resource: string;
  cursor_updated_since?: string | null;
  last_run_at?: string | null;
  last_success_at?: string | null;
  last_status?: string | null;
  last_error?: string | null;
  records_processed: number;
  is_running: boolean;
}

export interface SyncStatus {
  provider: string;
  healthy: boolean;
  resources: SyncState[];
}

export interface SyncRunResult {
  success: boolean;
  results: Record<
    string,
    {
      resource: string;
      processed: number;
      created: number;
      updated: number;
      skipped_locked: boolean;
      error?: string | null;
    }
  >;
}

// ---------- trainer ----------

export interface TrainerBatchSummary {
  id: number;
  external_id: string;
  name: string;
  code?: string | null;
  status?: string | null;
  mode?: string | null;
  location?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  student_count: number;
  program_id?: number | null;
  program_title?: string | null;
  module_count: number;
}

export interface TrainingSession {
  id: number;
  batch_id: number;
  batch_name?: string | null;
  module_id?: number | null;
  module_title?: string | null;
  title?: string | null;
  scheduled_start?: string | null;
  scheduled_end?: string | null;
  status: SessionStatus;
  join_code?: string | null;
  started_at?: string | null;
  ended_at?: string | null;
  materials_published_at?: string | null;
}

export interface TrainingSessionCreate {
  batch_id: number;
  module_id?: number | null;
  title?: string;
  scheduled_start?: string | null;
  scheduled_end?: string | null;
}

File: techpath-backend/app/schemas/training.py
"""Pydantic schemas for training content.

The interesting part is ``LectureAssetCreate``. ``lecture_assets`` is one wide table with
an ``asset_type`` discriminator, which keeps the storage simple but would leave the
payload unvalidated if we stopped there. So the shape of each type is pinned here
instead: a discriminated union means a quiz must carry questions, a YouTube asset must
carry a URL, and a markdown asset must carry a body — enforced on write, at the edge.
"""
from datetime import datetime
from typing import Annotated, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator

from app.core.constants import AssetStatus, AssetType, TrainingDeliveryMode


# ============================================
# Lecture assets — per-type payloads
# ============================================


class _AssetCommon(BaseModel):
    """Fields every asset carries regardless of type."""

    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    tags: Optional[List[str]] = None
    status: AssetStatus = AssetStatus.DRAFT


class InlineTextAssetIn(_AssetCommon):
    asset_type: Literal[
        AssetType.NOTES, AssetType.CHEAT_SHEET, AssetType.CODE_SNIPPET
    ]
    body: str = Field(..., min_length=1)
    # Only meaningful for code_snippet; harmless elsewhere.
    language: Optional[str] = Field(None, max_length=40)


class FileAssetIn(_AssetCommon):
    asset_type: Literal[
        AssetType.MARKDOWN,
        AssetType.PDF,
        AssetType.PPT,
        AssetType.VIDEO,
        AssetType.NOTEBOOK,
        AssetType.ZIP,
        AssetType.EXCEL,
        AssetType.CSV,
        AssetType.TERMINAL_RECORDING,
        AssetType.HTML_BUNDLE,
    ]
    media_file_id: int = Field(..., gt=0)


class LinkAssetIn(_AssetCommon):
    asset_type: Literal[AssetType.EXTERNAL_URL, AssetType.GITHUB_REPO, AssetType.YOUTUBE]
    external_url: HttpUrl


class QuizQuestion(BaseModel):
    question: str = Field(..., min_length=1)
    options: List[str] = Field(..., min_length=2)
    correct_index: int = Field(..., ge=0)
    explanation: Optional[str] = None

    @field_validator("correct_index")
    @classmethod
    def _index_in_range(cls, v: int, info) -> int:
        options = info.data.get("options")
        if options is not None and v >= len(options):
            raise ValueError(f"correct_index {v} is out of range for {len(options)} options")
        return v


class QuizAssetIn(_AssetCommon):
    asset_type: Literal[AssetType.QUIZ]
    questions: List[QuizQuestion] = Field(..., min_length=1)
    pass_mark_percent: int = Field(default=60, ge=0, le=100)


class AssignmentAssetIn(_AssetCommon):
    asset_type: Literal[AssetType.ASSIGNMENT]
    instructions: str = Field(..., min_length=1)
    due_in_days: Optional[int] = Field(None, ge=0)
    rubric: Optional[List[str]] = None


class LabStep(BaseModel):
    title: str = Field(..., min_length=1)
    instructions: str = ""


class LabAssetIn(_AssetCommon):
    asset_type: Literal[AssetType.LAB]
    objective: str = Field(..., min_length=1)
    steps: List[Union[LabStep, str]] = Field(..., min_length=1)
    starter_code: Optional[str] = None
    expected_output: Optional[str] = None


LectureAssetCreate = Annotated[
    Union[
        InlineTextAssetIn,
        FileAssetIn,
        LinkAssetIn,
        QuizAssetIn,
        AssignmentAssetIn,
        LabAssetIn,
    ],
    Field(discriminator="asset_type"),
]


class LectureAssetUpdate(BaseModel):
    """Partial update. The asset type is immutable — its payload column depends on it,
    so changing it would orphan the existing payload. Delete and recreate instead."""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    tags: Optional[List[str]] = None
    status: Optional[AssetStatus] = None
    is_active: Optional[bool] = None
    body: Optional[str] = None
    media_file_id: Optional[int] = None
    external_url: Optional[str] = None
    config: Optional[dict] = None


class CsvPreview(BaseModel):
    header: List[str] = []
    rows: List[List[str]] = []
    total_rows: int
    truncated: bool


class LectureAssetResponse(BaseModel):
    id: int
    public_id: str
    title: str
    asset_type: str
    description: Optional[str] = None
    body: Optional[str] = None
    media_file_id: Optional[int] = None
    external_url: Optional[str] = None
    config: Optional[dict] = None
    tags: List[str] = []
    status: str
    is_active: bool
    storage_kind: str
    # Resolved server-side from media_file_id via storage_service — local storage and
    # Azure Blob need different URL shapes (Azure's is a signed, expiring SAS URL), so
    # no client should ever construct this from a raw path itself.
    file_url: Optional[str] = None
    # Also resolved server-side, for the same reason file_url is: object storage
    # providers generally don't grant CORS to arbitrary origins by default, so a
    # browser-side fetch() of file_url would be silently blocked before it ever reaches
    # the network. A server-to-server fetch has no such restriction.
    csv_preview: Optional[CsvPreview] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AssetUsage(BaseModel):
    """Where an asset is currently placed. Shown before edit or delete so nobody
    changes shared content without seeing the blast radius."""

    module_id: int
    module_title: str
    program_id: int
    program_title: str

    model_config = ConfigDict(from_attributes=True)


# ============================================
# Modules
# ============================================


class TrainingModuleCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    description: Optional[str] = None
    display_order: int = 0
    estimated_minutes: Optional[int] = Field(None, ge=0)
    status: AssetStatus = AssetStatus.DRAFT


class TrainingModuleUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    slug: Optional[str] = Field(None, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    description: Optional[str] = None
    display_order: Optional[int] = None
    estimated_minutes: Optional[int] = Field(None, ge=0)
    status: Optional[AssetStatus] = None


class ModuleAssetLink(BaseModel):
    """An asset as placed in a module."""

    id: int
    asset_id: int
    display_order: int
    is_required: bool
    notes: Optional[str] = None
    asset: LectureAssetResponse

    model_config = ConfigDict(from_attributes=True)


class TrainingModuleResponse(BaseModel):
    id: int
    program_id: int
    title: str
    slug: str
    description: Optional[str] = None
    display_order: int
    estimated_minutes: Optional[int] = None
    status: str
    asset_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TrainingModuleDetail(TrainingModuleResponse):
    assets: List[ModuleAssetLink] = []


class AttachAssetRequest(BaseModel):
    asset_id: int = Field(..., gt=0)
    display_order: Optional[int] = None
    is_required: bool = True
    notes: Optional[str] = None


class ReorderItem(BaseModel):
    id: int = Field(..., gt=0)
    display_order: int = Field(..., ge=0)


class ReorderRequest(BaseModel):
    items: List[ReorderItem] = Field(..., min_length=1)


# ============================================
# Programmes
# ============================================


class TrainingProgramCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    summary: Optional[str] = Field(None, max_length=500)
    description: Optional[str] = None
    # Optional by design: offline-only training has no public course.
    course_id: Optional[int] = None
    delivery_mode: TrainingDeliveryMode = TrainingDeliveryMode.OFFLINE
    level: Optional[str] = Field(None, max_length=20)
    duration: Optional[str] = Field(None, max_length=100)
    cover_image: Optional[str] = Field(None, max_length=500)
    tags: Optional[List[str]] = None
    status: AssetStatus = AssetStatus.DRAFT
    display_order: int = 0


class TrainingProgramUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    slug: Optional[str] = Field(None, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    summary: Optional[str] = Field(None, max_length=500)
    description: Optional[str] = None
    course_id: Optional[int] = None
    delivery_mode: Optional[TrainingDeliveryMode] = None
    level: Optional[str] = None
    duration: Optional[str] = None
    cover_image: Optional[str] = None
    tags: Optional[List[str]] = None
    status: Optional[AssetStatus] = None
    display_order: Optional[int] = None
    is_active: Optional[bool] = None


class TrainingProgramResponse(BaseModel):
    id: int
    title: str
    slug: str
    summary: Optional[str] = None
    description: Optional[str] = None
    course_id: Optional[int] = None
    delivery_mode: str
    level: Optional[str] = None
    duration: Optional[str] = None
    cover_image: Optional[str] = None
    tags: List[str] = []
    status: str
    display_order: int
    is_active: bool
    module_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TrainingProgramDetail(TrainingProgramResponse):
    modules: List[TrainingModuleResponse] = []


class AssetTypeInfo(BaseModel):
    """Registry entry, published so the admin UI builds its type picker from the same
    source of truth the backend validates against."""

    value: str
    label: str
    storage_kind: str
    max_size_mb: int
    allowed_content_types: List[str]
    allowed_extensions: List[str]
